File: 240102/안전 지대/comfort-zone.cpp
#include <iostream>
#include <algorithm>
#define MAX_N 50
#define MAX_M 50
#define MAX_HEIGHT 100
#define DIR_NUM 4
using namespace std;

// 안전영역: K초과
int n, m;
int grid[MAX_N][MAX_N];
int ans=-1;     // 가능한 안전 영역의 최솟값이 0이므로 -1로 초기
int max_k;
bool visited[MAX_N][MAX_N];

// 탐색하는 위치가 격자 범위 내에 있는지 여부를 반환
bool InRange(int x, int y){
    return 0 <= x && x < n && 0 <= y && y < m;
}

// 탐색하는 위치로 움직일 수 있는지 여부 반
bool CanGo(int x, int y, int k){
    return InRange(x,y) && grid[x][y] > k && !visited[x][y];
}

// visited 배열을 초기화해주
void Initialize(){
    for(int i=0; i<n; i++){
        for(int j=0; j<m; j++){
            visited[i][j] = false;
        }
    }
}

void DFS(int x, int y, int k){
    int dirs[4][2] = {{1,0},{-1,0},{0,1},{0,-1}};
    
    // 네 방향 각각에 대하여 DFS 탐색을 수행
    for(int i=0; i<4; i++){
        int nx = x + dirs[i][0], ny = y + dirs[i][1];
        if(CanGo(nx,ny,k)){
            visited[nx][ny] = true;
            DFS(nx,ny,k);
        }
    }
}

int main() {
    int max_h = 0;
    cin >> n >> m;
    for(int i=0; i<n; i++){
        for(int j=0; j<m; j++){
            cin >> grid[i][j];
            max_h = max(max_h, grid[i][j]);
        }
    }

    // 각 가능한 비의 높이에 대하여 안전 영역의 수를 탐색하기
    for(int k=1; k<=max_h; k++){
        // 이 부분을 GetZoneNum으로 나누
        // 새로운 탐색을 시작한다는 의미로 cur(zone 개수)을 0으로 갱신
        // visited배열 초기화
        Initialize();
        int cur = 0;

        // 격자의 각 위치에 대하여 탐색을 시작할 수 있는 경우
        // 해당 위치로부터 시작한 DFS 탐색을 수행
        for(int i=0; i<n; i++){
            for(int j=0; j<m; j++){
                if(CanGo(i,j,k)){
                    // 해당 위치를 탐색할 수 있는 경우 visited 배열을 갱신하고
                    // 안전 영역을 하나 추가해주
                    visited[i][j] = true;
                    cur++;
                    DFS(i,j,k);
                }
            }
        }
        // 기존 최대 영역의 수(ans)보다 클(cur) 경우 이를 갱신하고 인덱스(k)를 저장
        if(ans < cur){
            ans = cur;
            max_k = k;
        }
    }

    cout << max_k << ' ' << ans;
    return 0;
}

File: 240127/안전 지대/comfort-zone.py
n, m = tuple(map(int, input().split()))
grid = [
    list(map(int, input().split()))
    for _ in range(n)
]
visited = [[False for _ in range(m)] for _ in range(n)]
dirs = ((1,0),(0,1),(-1,0),(0,-1))

def InRange(x,y):
    return 0 <= x and x < n and 0 <= y and y < m

def CanGo(x,y,val):
    return InRange(x,y) and not visited[x][y] and grid[x][y] > val 

def Initialize():
    global visited
    for i in range(n):
        for j in range(m):
            visited[i][j] = False

max_safe = 0
height = 0

def dfs(x,y,k):
    for d in range(4):
        nx, ny = x+dirs[d][0], y+dirs[d][1]
        if CanGo(nx,ny,k):
            visited[nx][ny] = True
            dfs(nx,ny,k)

for k in range(1,101):
    cur_safe = 0
    Initialize()
    for i in range(n):
        for j in range(m):
            if CanGo(i,j,k):
                cur_safe += 1
                visited[i][j] = True
                dfs(i,j,k)
    if max_safe < cur_safe:
        max_safe = cur_safe
        height = k

print(height, max_safe)